proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,324 @@
+
+We could propose to also store the new WALs in separate directories e.g. `wal.v2`. Thanks to that the rewrite from one version to another is **eventual** and can be done segment by segment without any forcible rewrite. The WAL will be rewritten within the next 2h of a normal operation. The additional advantage of this way of versioning is that it's clear when your WAL fully migrated to a certain version.
+
+Cons:
+* Semantics on reading mixed WAL situation it's more complex to get right.
+
+### Alternative: Record-based WAL versioning
+
+Given we usually change WAL by changing its records, the WAL version could be simply [max number of types](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/record/record.go#L54) we write to WAL.
+
+### Alternative: Maintain two WALs (well four, with WBL)
+
+Duplicating records would be too expensive for already painful latency and CPU/mem usage for e.g. replay.
+However, we could write an entirely new WAL with only a disk latency and space penalty, on top of writing old one, for compatibility.
+
+Cons:
+* Disk space increased
+* Inconsistent with the TSDB format strategy.
+* Complex to implement.
+
+### Alternative: Add a CLI tool that rewrites WAL to a specific version
+
+We could add a `promtool` command that rewrites WAL segments to a given version.
+
+Users can then migrate their WAL with a single command either as an init process before reverting or manually on a disk.
+
+This should take ~minutes even for larger WAL files.
+
+Pros:
+* Less need for a two-fold strategy?
+
+Cons:
+* We need to write migration code and ensure it's efficient enough for bigger data
+* A bit painful to use on scale and remotely (e.g. on Kubernetes)
+
+### Alternative: Rewrite WAL before/during replay
+
+Instead of supporting multiple directories for WAL for various versions, we could rewrite WAL on the start.
+
+Pros:
+* Simpler implementation
+
+Cons:
+* We already suffer from replay problems, so I propose an eventual rewrite.
+* Rewrite is more risking than read-only WAL (of previous version)
+
+### Alternative: Don't version WAL, don't introduce a flag
+
+Pro:
+* One issue with a flag is that we lose "automatic" rollout for some users e.g. what if we will have another v3 WAL version, but users sets an explicit `--storage.tsdb.stateful.write-wal-version=v3` option? For those users they will need to make a manual choice and there might be not aware of benefits of a new WAL or do any work to test new WAL for their setup. With raw two-fold strategy format is forced, but also transparent (if without issues).
+
+Cons:
+* Takes more time for features to be available to users
+* Demotivating for format changes (long feedback loop)
+* Harder to communicate what exactly changed in each Prometheus version or even implement backward compatibility?
+
+### Alternative: Use feature flag instead
```

Comment:
I reviewed the linked issues and efforts motivating the segments versioning, and I believe all of them can be managed with the two-fold strategy/migration and a feature flag, so maybe I'm missing sth or maybe more concrete scenarios are missing in the proposal.

Unless we intend to change the record's header format (which I think we can also do without having to verson the WAL segments), I think the current format is generic enough to handle all other changes. It's a map of `record type/code -> how to handle it`, I think we can easily customize that mapping (and improve it if needed) without having to version the segments.

For example, If I were to propose an optimized encoding for samples, I would add a `SampleOptimized` record type (with its reader/writer) and a `--feature-flag=optimized-wal-encoding` or sth to trigger reading and then writing in another version. With the two-fold strategy and proper documentation/assistance on potential downgrades [1], as suggested above, and on feature gate deactivation [2], I believe we could roll out and test that `SampleOptimized`. This approach aligns with how we've proceeded so far, as you mentioned.

While versioning WAL segments provides more visibility into the version in use, what does a `v3` segment version really mean to a user? Does it mean native histograms are supported? Does it mean optimized encoding is enabled? I think it would only add confusion.

Also, having such vague versions would create additional confusion and necessitate more synchronization between developers working concurrently on changes that require "incrementing the WAL segment version".

---

[1] In addition to the documentation, we could implement a built-in safeguard mechanism for downgrades: store the minimum downgrade version on disk and refuse to start up unless the file is manually deleted or another action is taken. 
Also, if we want to raise awareness about the "dangers of unsupported downgrading", we may need to reconsider how `record.Unknown` is handled.

[2] We could leverage this proposal to improve the discoverability of built-in feature gates (I think there is an issue for this somewhere).

Reply:
Thanks, fair points. Sounds like you propose a feature flags with feature names, which semantically might be close to "record versioning" instead of a single "segment version" proposed right now. It seems the main difference is organizing features into version umbrella or deliver them in separation. Definitely something to consider, let's start unpacking pros & cons.

> While versioning WAL segments provides more visibility into the version in use, what does a v3 segment version really mean to a user? Does it mean native histograms are supported? Does it mean optimized encoding is enabled? I think it would only add confusion. Also, having such vague versions would create additional confusion and necessitate more synchronization between developers working concurrently on changes that require "incrementing the WAL segment version".

Yes. It will means immediately nothing what it changes to user and that's on purpose. This versioning is optimized for telling users about important data compatibility and discovering what version of Prometheus supports what versions of WAL, that's it.

Imagine we change WAL and finally use proper `nativehistogram` records (instead of 4 different records now) as a `wal-nhcleanup` feature, then new sample with CT with `wal-ctsample` feature, finally let's say structured summary in future with `wal-summary` feature. Now you deploy new Prometheus version that (after a fold) defaults to `wal-nhcleanup`, then user manually opt-in into `wal-ctsample`  and  `wal-summary`  is still opt-out. Let's say user reverts to the previous Prometheus release it was running (e.g. LTS or let's say 3 versions back) and some data are completely missing because of WAL incompatibility. Finding what feature actually is incompatible and what version of Prometheus supports what features is not easy. Talking in segment versions is a bit more consistent and obvious when `ls`-ing your directory or logging on unknown segment version.

But.. maybe that user experience is not needed. Maybe asking those tricky features to be in LTS version before switching to opt-in would help (I would avoid having to wait for v4.x Prometheus etc), better docs could be added too.

Reply:
> and some data are completely missing because of WAL incompatibility

As I mentioned earlier, segment versioning isn't a perfect solution for this issue. If a user downgrades to a Prometheus version that can't decode `v2` segments, the data within those segments will be lost. However, if users follow the recommended downgrade paths, the feature flag approach should be able to handle those scenarios without data loss.

> Finding what feature actually is incompatible and what version of Prometheus supports what features is not easy. Talking in segment versions is a bit more consistent and obvious when ls-ing your directory or logging on unknown segment version.

I think adding a new dimension (segment version) to the mix would just add to the confusion here, especially since users will still need to consult documentation of other tools to figure out the differences between segment `v3` and `v2`.

I think well-documented feature flags and a well-documented compatibility matrix should be enough. The most important thing, as you mentioned, is to provide a well-tested fallback (through the two-fold strategy).

So, what I think we could still do through this proposal is:

- Officially document the two-fold strategy, with examples and schemas.
- Start requiring such "invasive/critical" (or maybe all of them)  feature flags to:
    - Ensure a fallback/LTS version exists and is mentioned in the docs before shipping the "invasive/feature flag in write mode".
    - Explicitly document what happens when the feature flag is disabled in the relevant versions.
- Add the needed utils/infra to allow e2e tests concerning those scenarios, in addition to those e2e tests, the feature flag can always be tested on long running stage envs even though a fallback version isn't ready yet, as long as the user is aware of that...